common/authorization/oauthAuthorizer.go

```diff
@@ -0,0 +1,122 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package authorization
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/cristalhq/jwt/v3"
+	log "github.com/sirupsen/logrus"
+	"go.uber.org/yarpc"
+
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/config"
+	"github.com/uber/cadence/common/log/tag"
+)
+
+type oauthAuthority struct {
+	authorizationCfg config.OAuthAuthorizer
+}
+
+type jwtClaims struct {
+	Sub        string
+	Name       string
+	Permission string
+	Domain     string
+	Iat        int64
+	TTL        int64
+}
+
+// NewOAuthAuthorizer creates a oauth authority
+func NewOAuthAuthorizer(
+	authorizationCfg config.OAuthAuthorizer,
+) Authorizer {
+	return &oauthAuthority{
+		authorizationCfg: authorizationCfg,
+	}
+}
+
+// Authorize defines the logic to verify get claims from token
+func (a *oauthAuthority) Authorize(
+	ctx context.Context,
+	attributes *Attributes,
+) (Result, error) {
+	call := yarpc.CallFromContext(ctx)
+	verifier, err := a.getVerifier()
+	if err != nil {
+		return Result{Decision: DecisionDeny}, err
+	}
+	token := call.Header(common.AuthorizationTokenHeaderName)
+	claims, err := a.parseToken(token, verifier)
+	if err != nil {
+		log.Debug("request is not authorized", tag.Error(err))
+		return Result{Decision: DecisionDeny}, nil
+	}
+	validationErr := a.validateClaims(claims, attributes)
+	if validationErr != nil {
+		log.Debug("request is not authorized", tag.Error(err))
+		return Result{Decision: DecisionDeny}, nil
+	}
+	return Result{Decision: DecisionAllow}, nil
+}
+
+func (a *oauthAuthority) getVerifier() (jwt.Verifier, error) {
+	publicKey, err := common.StringToRSAPublicKey(a.authorizationCfg.JwtCredentials.PublicKey)
+	if err != nil {
+		return nil, err
+	}
+	algorithm := jwt.Algorithm(a.authorizationCfg.JwtCredentials.Algorithm)
+	verifier, err := jwt.NewVerifierRS(algorithm, publicKey)
+	if err != nil {
+		return nil, err
+	}
+	return verifier, nil
+}
+
+func (a *oauthAuthority) parseToken(tokenStr string, verifier jwt.Verifier) (*jwtClaims, error) {
+	token, verifyErr := jwt.ParseAndVerifyString(tokenStr, verifier)
+	if verifyErr != nil {
+		return nil, verifyErr
+	}
+	var claims jwtClaims
+	_ = json.Unmarshal(token.RawClaims(), &claims)
+	return &claims, nil
+}
+
+func (a *oauthAuthority) validateClaims(claims *jwtClaims, attributes *Attributes) error {
+	if claims.TTL > a.authorizationCfg.MaxJwtTTL {
```

Comment:
@longquanzheng is this how you wanted me to validate `maxjwtTTL`? if somehow maxjwtTTL changes it might potentially break all the tokens, but I wonder if this is expected.

Reply:
I think that’s expected. The maintainer should know what they are doing when changing the config.